### PR TITLE
openstack: allow external nameserver to access security groups

### PR DESCRIPTION
**Pull request by smithfarm**
Signed-off-by: Nathan Cutler <ncutler@suse.com>

### CODE REVIEW

File: teuthology/openstack/__init__.py [hunk 1]
```diff
@@ -1159,6 +1163,13 @@ def create_security_group(self):
 openstack security group rule create --src-group {server} --dst-port 1:65535 {server}
 openstack security group rule create --protocol udp --src-group {server} --dst-port 1:65535 {server}
         """.format(server=self.server_group(), worker=self.worker_group()))
+        if self.args.nameserver:
+            misc.sh("""
+openstack security group rule create --protocol tcp --src-ip {nameserver} --dst-port 53 {server} # for DNS access
+openstack security group rule create --protocol udp --src-ip {nameserver} --dst-port 53 {server} # for DNS access
+openstack security group rule create --protocol tcp --src-ip {nameserver} --dst-port 53 {worker} # for DNS access
+openstack security group rule create --protocol udp --src-ip {nameserver} --dst-port 53 {worker} # for DNS access
```
Comment (jan--f) [inline teuthology/openstack/__init__.py]: I would think it's enough to give the target nodes access to port 53 of the nameserver. Can't see a scenario where the nameserver will try to connect to target nodes.

File: teuthology/openstack/__init__.py [hunk 2]
```diff
@@ -1062,6 +1062,10 @@ def get_user_data(self):
         if self.args.simultaneous_jobs:
             all_options += [ '--nworkers ' + str(self.args.simultaneous_jobs) ]
         if self.args.nameserver:
+            try:
+                socket.inet_aton(self.args.nameserver)
+            except socket.error:
+                raise Exception("nameserver must be a valid IP address")
```
Comment (kshtsk) [inline teuthology/openstack/__init__.py]: 1) This change is not related to the PR.
2) In general name server should not be a valid IP only.

File: teuthology/openstack/__init__.py [hunk 3]
```diff
@@ -1159,6 +1163,13 @@ def create_security_group(self):
 openstack security group rule create --src-group {server} --dst-port 1:65535 {server}
 openstack security group rule create --protocol udp --src-group {server} --dst-port 1:65535 {server}
         """.format(server=self.server_group(), worker=self.worker_group()))
+        if self.args.nameserver:
```
[the remark this answers is not in the document]
Comment (smithfarm) [reply teuthology/openstack/__init__.py]: The way I see this, it is allowing traffic to ports 1:65535 on the teuthology server only.

File: teuthology/openstack/__init__.py (hunk 2 above)
Comment (smithfarm) [reply teuthology/openstack/__init__.py]: Wouldn't it be crazy to specify the nameserver as a hostname, though?

Would you specify a hostname for the nameserver in /etc/resolv.conf ?

File: teuthology/openstack/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kshtsk) [reply teuthology/openstack/__init__.py]: Yes, but there is the same for worker.

[the remark this answers is not in the document]
Comment (kshtsk) [reply teuthology/openstack/__init__.py]: Hm... sorry, you're right, have you tested this code on OVH?

[the remark this answers is not in the document]
Comment (smithfarm) [reply teuthology/openstack/__init__.py]: no, this is just an idea

File: teuthology/openstack/__init__.py (hunk 2 above)
Comment (kshtsk) [reply teuthology/openstack/__init__.py]: Get hostname ip before you need to set IP?

**Comment by smithfarm** [comment]
So, after discussing this with @jan--f it would seem no additional security rules are needed for the targets ("workers") to be able to use the external nameserver, assuming the external nameserver does not block the incoming traffic in some way - i.e. it should allow TCP and UDP traffic on port 53 from the targets.

And I verified this in a teuthology target machine - I can do e.g.:

```
target158069066172:/home/ubuntu # host www.harvard.edu 8.8.8.8
Using domain server:
Name: 8.8.8.8
Address: 8.8.8.8#53
Aliases: 

www.harvard.edu is an alias for www.harvard.edu.cdn.cloudflare.net.
www.harvard.edu.cdn.cloudflare.net has address 104.16.152.6
www.harvard.edu.cdn.cloudflare.net has address 104.16.151.6
www.harvard.edu.cdn.cloudflare.net has address 104.16.153.6
www.harvard.edu.cdn.cloudflare.net has address 104.16.155.6
www.harvard.edu.cdn.cloudflare.net has address 104.16.154.6
target158069066172:/home/ubuntu # host www.harvard.edu ceph-ci.suse.de
Using domain server:
Name: ceph-ci.suse.de
Address: 158.69.90.90#53
Aliases: 

www.harvard.edu is an alias for www.harvard.edu.cdn.cloudflare.net.
www.harvard.edu.cdn.cloudflare.net has address 104.16.152.6
www.harvard.edu.cdn.cloudflare.net has address 104.16.154.6
www.harvard.edu.cdn.cloudflare.net has address 104.16.153.6
www.harvard.edu.cdn.cloudflare.net has address 104.16.151.6
www.harvard.edu.cdn.cloudflare.net has address 104.16.155.6
```

File: teuthology/openstack/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (kshtsk) [reply teuthology/openstack/__init__.py]:
```openstack security group rule create --src-group {worker} --dst-port 1:65535 {worker}
openstack security group rule create --protocol udp --src-group {worker} --dst-port 1:65535 {worker}
```
What does this line mean from your point?

[the remark this answers is not in the document]
Comment (kshtsk) [reply teuthology/openstack/__init__.py]: > The way I see this, it is allowing traffic to ports 1:65535 on the teuthology server only.

Not the server, but the server group.

**Comment by kshtsk** [comment]
> allow external nameserver to access security groups 
more general question, why should we allow external nameserver to access workers?

**Comment by smithfarm** [comment]
This was a misguided attempt to address https://github.com/SUSE/DeepSea/issues/1438 which turned out to be caused by a misconfiguration on the external nameserver.
